src/cobra/core/gene.py

```diff
@@ -299,3 +271,360 @@ def _repr_html_(self):
+
+    @classmethod
+    def from_string(cls: type, string_gpr: str) -> "GPR":
+        """Construct a GPR from a string.
+
+        Parameters
+        ----------
+        string_gpr: str
+            a string that describes the gene rules, in a format like
+            A & B
+
+        Returns
+        -------
+        GPR:
+            returns a new GPR while setting  self.body as
+            Parsed AST tree that has the gene rules
+            This function also sets self._genes with the gene ids in the AST
+
+        """
+        gpr = cls()
+        if not isinstance(string_gpr, str):
+            raise TypeError(f"GPR().from_string requires a str, not {type(str)}")
+        uppercase_AND = re.compile(r"\bAND\b")
+        uppercase_OR = re.compile(r"\bOR\b")
+        str_expr = string_gpr.strip()
+        if len(str_expr) == 0:
+            gpr.body = []
+            return gpr
+        for char, escaped in replacements:
+            if char in str_expr:
+                str_expr = str_expr.replace(char, escaped)
+        escaped_str = keyword_re.sub("__cobra_escape__", str_expr)
+        escaped_str = number_start_re.sub("__cobra_escape__", escaped_str)
+
+        try:
+            tree = ast_parse(escaped_str, "<string>", "eval")
+        except (SyntaxError, TypeError) as e:
+            if "AND" in string_gpr or "OR" in string_gpr:
+                warn(
+                    "uppercase AND/OR found in rule '%s' for '%s'"
+                    % (string_gpr, repr(gpr)),
+                    SyntaxWarning,
+                )
+                string_gpr = uppercase_AND.sub("and", string_gpr)
+                string_gpr = uppercase_OR.sub("or", string_gpr)
+            try:
+                tree = ast_parse(string_gpr, "<string>", "eval")
+            except SyntaxError as e:
+                warn(
+                    "malformed gene_reaction_rule '%s' for %s"
+                    % (string_gpr, repr(gpr)),
+                    SyntaxWarning,
+                )
+                warn("GPR will be empty")
+                warn(e.msg)
+                return gpr
+        return cls(tree)
+
+    @property
+    def genes(self) -> FrozenSet:
+        """To check the genes.
+
+        This property updates the genes before returning them, in case the GPR was
+        changed and the genes weren't.
+
+        Returns
+        -------
+        genes: frozenset
+            All the genes in a frozen set. Do not try to change them with this property.
+        """
+        self.update_genes()
+        return frozenset(self._genes)
+
+    def update_genes(self) -> None:
+        """Update genes, used after changes in GPR.
+
+        Walks along the AST tree of the GPR class, and modifies self._genes
+
+        """
+        if hasattr(self, "body"):
+            walker = GPRWalker()
+            walker.visit(self)
+            self._genes = walker.gene_set
+            if "" in self._genes:
+                self._genes.remove("")
+
+    def __eval_gpr(
```

Comment:
Double underscore methods can be nasty when it comes to class inheritance. Nothing is truly private in Python anyway so I think a single underscore is acceptable.

```suggestion
    def _eval_gpr(
```